src/Nri/Ui/MediaQuery/V2.elm

```diff
@@ -0,0 +1,379 @@
+                        internal.breakpoints
+            }
+
+        MediaQuery (UserPreference onQuery offQuery) satisfiesTargetStyles doesNotSatisfyTargetStyles ->
+            { internal
+                | userPreferences =
+                    internal.userPreferences
+                        |> Dict.update onQuery (merge satisfiesTargetStyles)
+                        |> Dict.update offQuery (merge doesNotSatisfyTargetStyles)
+            }
+
+
+{-| Build a list of `Css.Style` from a ResponsiveStyles record.
+
+    Styles are output in the following order:
+
+    1. User preferences
+    2. Mobile-first breakpoints (min-width, ascending)
+    3. Desktop-first breakpoints (max-width, descending)
+
+    This ordering (at least for the breakpoints) is important to ensure the correct styles are applied.
+
+    CASCADES, BABY!
+
+-}
+toStyles : ResponsiveStyles -> List Style
+toStyles { breakpoints, userPreferences } =
+    let
+        prepend =
+            Maybe.cons
+
+        append maybeItem list =
+            Maybe.map (List.singleton >> List.append list) maybeItem |> Maybe.withDefault list
+
+        mkUserPreferenceQuery ( preference, styles ) =
+            withMediaQuery [ preference ] styles
+
+        mkViewportQuery rule px =
+            Maybe.map <| withMedia [ only screen [ rule <| Css.px px ] ]
+
+        addViewportQuery ( px, ( desktopFirst, mobileFirst ) ) =
+            prepend (mkViewportQuery maxWidth px desktopFirst) >> append (mkViewportQuery minWidth (px + 1) mobileFirst)
+    in
+    List.map mkUserPreferenceQuery (Dict.toList userPreferences)
+        ++ List.foldr addViewportQuery [] (Dict.toList breakpoints |> List.sortBy Tuple.first)
+
+
+{-| Build a single `Css.Style` from a ResponsiveStyles record.
+
+    Styles are output in the following order:
+
+    1. User preferences
+    2. Mobile-first breakpoints (min-width, ascending)
+    3. Desktop-first breakpoints (max-width, descending)
+
+    This ordering (at least for the breakpoints) is important to ensure the correct styles are applied.
+
+    This is a convenience function for `Css.batch (toStyles queries)`.
+
+-}
+toStyle : ResponsiveStyles -> Style
+toStyle =
+    toStyles >> Css.batch
+
+
+{-| Build a single `Css.Style` from a list of media queries.
+
+    !!!  ARE YOU TRYING TO USE THIS FUNCTION AND GETTING A TYPE ERROR ON YOUR LIST OF MEDIA QUERIES? READ THIS! !!!
+
+    This module is designed with a pipeline-style API in mind, and uses phantom types to enforce certain restrictions.
+
+    These phantom types prevent using modifiers like `not` and `offset` with media queries that are not designed to accept them.
+
+    This function is an alias for `List.foldl on init >> toStyle` provided for convenience and the common case.
+
+    If you're stuck, you have 2 options:
+
+    1. Use 2 independent calls to `MediaQuery.fromList`, one with your breakpoints and one with your user preferences.
+
+    2. Use the pipeline-style API directly.
```

Comment:
Is there a benefit from one patter over the other? I would rather have a single api if possible

Reply:
Pipeline-style API is more flexible and lets us use phantom types for media query constructors–e.g. we can implement `offset` and only allow it to be used with breakpoints, since we can't offset a user preference.

The problem arises if you have 2 `MediaQuery attribute` with disparate attributes. Consider the following:

```elm
MediaQuery.init
  |> MediaQuery.on MediaQuery.mobile []
  |> MediaQuery.on MediaQuery.prefersReducedMotion []
  |> MediaQuery.toStyle
```

This works, because `on` interprets the type of the `MediaQuery properties`, does the right thing for that media query type, and returns a normalized `ResponsiveStyles` record. Since `ResponsiveStyles` is unbound, we can repeat this operation with a bunch of media queries without caring about the phantom types. 

Now, consider the List-style API:

```elm
MediaQuery.fromList
  [ MediaQuery.mobile []
  , MediaQuery.prefersReducedMotion []
  ]
```

What is the type? `MediaQuery { offsettable : () }`? or `MediaQuery {}`.

Trick question, the answer is compiler error.

For end-consumers in practice since user preferences and breakpoints can't conflict in the cascade, this doesn't really matter. It will throw a potentially confusing compiler error, but can be resolved by using

```
[ MediaQuery.fromList [MediaQuery.mobile []]
, MediaQuery.fromList [MediaQuery.prefersReducedMotion []]
```

It is however an actual problem for any libraries that want to implement this. Consider [this line](https://github.com/NoRedInk/noredink-ui/pull/1659/files#diff-69b280c9b764b0a83df1dbec0ed407a0cfab0f5b56346abec0bd7c2c19030e2eR129) in Tooltip that defines the responsive types on the config.

```elm
type alias Tooltip msg =
    { responsiveTooltipStyleOverrides : MediaQuery.ResponsiveStyles }
```

If you have this implemented/exposed strictly as a list-style API, you're in the same issue as above where you're restricted to media queries that share the same phantom type.

---

Pros of the list-style API:
  - It covers most cases for general consumers that are applying the results directly to some HTML, not making re-usable components.
  - It's more consise.
  - It better matches existing Nri.Ui behavior*

*I have thoughts about this, given all of the above about the implications for phantom types and using combinators instead of a ton of permutations of attributes, and think broader adoption of this style api may actually be a good thing.